Password Safe Time-based one-time password (TOTP)

AshleyT3 opened the pull request:
Password Safe Time-based one-time password (TOTP)

AshleyT3 commented:
@ronys These are initial TOTP commits adding SHA1/HMAC-SHA1 tests. PR created in "draft" mode at this time.

AshleyT3 commented:
@ronys commit [8b45b09](https://github.com/pwsafe/pwsafe/pull/1035/commits/8b45b099d4fab6368e5a23069fe84de10918a26c) is a proposed TOTP Parameters entry field.

AshleyT3 commented:
@ronys moved TOTP Parameters entry field from V4 to V3, see commit [dd99a51](https://github.com/pwsafe/pwsafe/pull/1035/commits/dd99a5121e863ea96282038ad9b87bb0c568a795)

AshleyT3 commented:
@ronys There are two proposed variations for V3 DB schema field additions. The first packs multiple values into a single field, the second (just pushed) uses separate fields, is simpler, easier to maintain as managed by existing internals around DB field handling. The first, with packed fields, was to conserve field ID space... not sure if you'd weigh that as worthy, all considering.

- Variation 1, with packed TOTP parameters: [commit dd99a51](https://github.com/pwsafe/pwsafe/commit/dd99a5121e863ea96282038ad9b87bb0c568a795)
- Variation 2, with separate TOTP parameter fields: [commit 4c6fdc8](https://github.com/pwsafe/pwsafe/pull/1035/commits/4c6fdc86b06ff9598747e2c285adc46205a7dbb2)

(will squash once solidified.)

ronys commented on docs/formatV3.txt:
```diff
@@ -341,6 +341,10 @@ Credit Card Expiration      0x1d        Text          N              [24]
 Credit Card PIN             0x1f        Text          N              [24]
 QR Code                     0x20        Text          N              [25]
+TOTP Config                 0x21        1 byte        N              [29]
+TOTP Digits                 0x22        1 byte        N              [29]
```
I think "TOTP Length" would be clearer, no?

ronys replied:
In the RFC, there's a section (5.1) defining "Digit" as the "number of digits in a HOTP value [...]".
Since we don't have such as section, I vote for clarity, that is, "TOTP Length".
If we really want to be pedantic, we can note that this corresponds to the term "Digits" in the aforementioned RFC.

ronys commented on docs/formatV3.txt:
```diff
@@ -530,6 +534,33 @@ to QR code without any further encoding.
 [28] An explicit end of entry field is useful for supporting new fields
 without breaking backwards compatability.
 
+[29] The TOTP parameters fields are TOTP Config (0x21), TOTP Digits (0x22),
```
s/Digits/Length/

ronys commented on docs/formatV3.txt:
```diff
@@ -530,6 +534,33 @@ to QR code without any further encoding.
 
+[29] The TOTP parameters fields are TOTP Config (0x21), TOTP Digits (0x22),
+TOTP Time Step (0x23), and TOTP Start Time (0x24). If any of these fields is not
+present, a default applies as outlined below. These fields, or their implicit
```
The defaults are "specified" or "defined" below, not "outlined".

ronys commented on docs/formatV3.txt:
```diff
@@ -530,6 +534,33 @@ to QR code without any further encoding.
+                      If this TOTP Config field is not present, the default
+                      HMAC hash algorithm is SHA1 (HMAC-SHA1).
+
+    TOTP Digits     = 1 byte indiating the number of TOTP code digits.
```
s/Digits/Length/

ronys commented on docs/formatV3.txt:
```diff
@@ -530,6 +534,33 @@ to QR code without any further encoding.
+                      This is the RFC4226 "Digit" value. If not specified,
+                      the default is 6 digits (0x06).
+
+    TOTP Time Step  = 1 byte indicating the RFC6238 time step "X" value.
```
"...value, in seconds.

ronys commented on src/core/crypto/RFC3548_Base32Decoder.h:
```diff
@@ -0,0 +1,90 @@
+  }
+
+  RFC3548_Base32Decoder& operator=(const RFC3548_Base32Decoder& other) {
+    clear();
```
What about the standard idiom of checking if &other != this?

ronys commented on src/core/crypto/RFC3548_Base32Decoder.h:
```diff
@@ -0,0 +1,90 @@
+  }
+
+  RFC3548_Base32Decoder& operator=(RFC3548_Base32Decoder&& other) noexcept {
+    clear();
```
see previous comment

ronys commented on src/core/crypto/external/Chromium/original_base32.c:
Do we need this file, or is it temporarily here?

ronys replied:
Please remove them and refer to URL and retrieval date.

AshleyT3 replied:
ACK, PR updated.

ronys commented on src/core/crypto/external/Chromium/original_base32.h:
ditto.

ronys commented on src/test/Base32Test.cpp:
```diff
@@ -0,0 +1,110 @@
+
+// The odd string splits below are an effort to mitigate
+// CI/CD false positives around secrets detection.
+vector<tuple<const char*, const char*>> base32_test_cases{
```
What's the source of these test vectors? Worth stating, perhaps?

AshleyT3 replied:
PR updated w/test vectors including both non-printable and RFC4648 Test Vectors.

ronys commented on src/test/HMAC_SHA1Test.cpp:
Are there standard test vectors for this?

AshleyT3 replied:
ACK, RFC2202 has HMAC-SHA1 cases using strings close to these but w/slight differences so let me just grab those.

AshleyT3 replied:
PR updated.

ronys commented on src/test/SHA1Test.cpp:
Ditto.

AshleyT3 replied:
ACK, let me look into it.

ronys commented on src/core/CoreOtherDB.cpp:
Doesn't the comment describing the difference flags need to be updated?

ronys commented on src/core/ItemData.h:
```diff
@@ -84,12 +85,42 @@ class CItemData : public CItem
   // Convenience: Get the untranslated (English) name of a FieldType
   static stringT EngFieldName(FieldType ft);
 
+  // Return XML-complient element name.
```
s/complient/compliant/

ronys commented on src/core/ItemData.h:
```diff
@@ -84,12 +85,42 @@ class CItemData : public CItem
+  size_t GetTwoFactorKeyLength() const { return GetField(TWOFACTORKEY).length(); }
+
+  uint8_t GetTotpConfigAsByte() const { return GetFieldAsByte(TOTPCONFIG, TOTP_CONFIG_ALGORITHM_DEFAULT); }
+  StringX GetTotpConfig() const { return IntegralToStringX(GetTotpConfigAsByte()); }
```
Shouldn't "Integral" be "Integer"?

ronys commented on src/core/TotpCore.cpp:
```diff
@@ -0,0 +1,71 @@
+* http://www.opensource.org/licenses/artistic-license-2.0.php
+* Contributed by Ashley R. Thomas, 18-Oct-2023
+*/
+// TotpCore.h
```
/.h/.cpp/

ronys commented on src/core/TotpCore.h:
Would be nice to wrap these into a namespace, e.g., PWSTotp?

ronys commented on src/core/TotpCore.h:
```diff
@@ -0,0 +1,35 @@
+ * funtionality connecting PwSafe Core users (i.e., UI/CLI) with TOTP internals.
+ */
+
+const uint32_t TOTP_INVALID_AUTH_CODE = 0;
```
An enum with relevant errors would be nicer, I think.

ronys commented on src/core/Util.h:
```diff
@@ -319,6 +326,23 @@ bool FindNoCase( const StringX& src, const StringX& dest);
+}
+
+template<class T>
+StringX IntegralToStringX(T v) {
```
IntegerToStringX ?

ronys commented on src/core/Util.h:
```diff
@@ -319,6 +326,23 @@ bool FindNoCase( const StringX& src, const StringX& dest);
+template<class T>
+StringX IntegralToStringX(T v) {
+  // For consistency for all integers, and specifically forcing single byte
+  // integrals (char, uint8_t, etc.) to be converted as integers to string
```
integers

ronys commented on src/test/TestCommon.h:
```diff
@@ -0,0 +1,41 @@
+// TestCommon.h
+//-----------------------------------------------------------------------------
+
+#pragma once
```
This isn't portable. Please wrap the contents in the standard
#ifndef _FILENAME_H
#define _FILENAME_H
...
#endif
instead

AshleyT3 commented:
@ronys There are still mysterious "builds.sr.ht" build errors... per discussions, I think you need to fix those... it seems caused by an expired token/creds for a dependent "builds.sr.ht" service. Thanks.

"builds.sr.ht ... failed to submit sr.ht job: gqlclient: HTTP server error (403 Forbidden): gqlclient: server failure: Authentication error: Invalid or expire..."

![image](https://github.com/pwsafe/pwsafe/assets/32182932/6ea2f151-9423-4b0e-861d-0bb60e70e002)

AshleyT3 reviewed (commented):
@ronys see comment at the following link which replaced this comment...
https://github.com/pwsafe/pwsafe/pull/1035#pullrequestreview-1732491049

AshleyT3 reviewed (commented):
@ronys As requested, I added the TOTP secret key field to the "Additional" tab. **Can you look, sign off?**

Following relates to UI changes in commit [f88b2e2](https://github.com/pwsafe/pwsafe/pull/1035/commits/f88b2e2f57b70df4137c3f83534fb61e5e9a2187)

Since all of these property pages are somewhat related in that they share the same runtime size, the ability to effectively apply/assess design changes, etc., are negatively affected if their design-time sizes are mismatched. I therefore aligned their design-time sizes in this change. This will help with TOTP UI changes and others going forward (so long as they are maintained to match the largest).

There were some minor nits such as missing or non-functioning accelerators, as well as some alignment/margin issues tweaked as well. This helps look at each page, assess functionality, alignment/margin for parity and so on.

Note, there are additional cleanup changes that could be applied... these were minimal to have a good view of the overall UI as we add fields for the set of TOTP UI changes.

A separate future "UI cleanup" PR can be used to perform other cleanup changes. For example, there is large vertical gapping between some controls which could likely be reduced to buy some UI real estate, or be more compact/leaner to match control sizes, if desired. 

In the old rc (before this change), some of the resources were not in property page order, so aligning them should make comparing in github easier going forward (so long as order is maintained).

To help review this change which aligns that ordering, you might want to do what I do, which is to copy/paste the property page resources for the Add/Edit property changes from old/new .rc file into separate "old" and "new" text files, putting the resources in prop sheet order... the following order...

IDD_ADDEDIT_BASIC DIALOGEX
IDD_ADDEDIT_ADDITIONAL DIALOGEX
IDD_ADDEDIT_DATETIMES DIALOGEX
IDD_ADDEDIT_PASSWORDPOLICY DIALOGEX
IDD_ADDEDIT_ATT DIALOGEX

IDD_ADDEDIT_BASIC_SHORT DIALOGEX
IDD_ADDEDIT_ADDITIONAL_SHORT DIALOGEX
IDD_ADDEDIT_DATETIMES_SHORT DIALOGEX
IDD_ADDEDIT_PASSWORDPOLICY_SHORT DIALOGEX
IDD_ADDEDIT_ATT_SHORT DIALOGEX

...after that, you can compare those text files in your favorite diff program.

For convenience, I've attached a ZIP which shows the result of what I just explained. Within are old/new rc and an html printout of a comparison report (as always, please scan or use zip safely)...

[cmp-old-vs-new-pwsafe-rc.zip](https://github.com/pwsafe/pwsafe/files/13367218/cmp-old-vs-new-pwsafe-rc.zip)

AshleyT3 commented:
@ronys I added the TOTP shared secret field to the Additional tab per discussions, and here are the screenshots you wanted to see for review. Can you review and let me know if you have any changes or if we are good to go with this placement? We can tweak later, I just want to ensure the general placement/location for normal/wide dialogs on Additional tab, and anything else passes your review. 

(Note, these are taken from the dialog editor so do not have icons/bitmaps loaded... but it's the new "Password Safe TOTP Authenticator" field you wanted on the Additional tab that I'm asking about.

Dialog IDD_ADDEDIT_ADDITIONAL:

![image](https://github.com/pwsafe/pwsafe/assets/32182932/8e58727e-c74b-489f-9c17-12cc699f88c1)

Dialog IDD_ADDEDIT_ADDITIONAL_SHORT:

![image](https://github.com/pwsafe/pwsafe/assets/32182932/75f4ef57-13f4-4e54-aadb-6c8debbd0445)

AshleyT3 commented:
@ronys (gentle ping) ... see above... screenshots you requested... can you review? Should be quick, just one field added. Thanks.

ronys commented:
Some alignment problems have crept in:
- Width of the action dropdowns has shrunk, they are no longer aligned with the other edit boxes in the first variant.
- "Keyboard Shortcut" has been truncated to "Keyboard" on both variants.
- "Double-Click Action:" and "Shift Double-Click Action:" have likewise been truncated in the first variant.
- I didn't build and test the tab order, I'd expect the TOTP key field to be tabbed into after keyboard shortcut

Two thoughts on the "Password Safe TOTP Authenticator" text:
1. "Password Safe" is redundant, as this is in the application. I don't think we self-reference anywhere else
2. "TOTP Authenticator", while accurate, isn't very common terminology, in the few apps that I've seen.
Suggest replacing this with "Authenticator Secret" and a (?) tooltip "The secret key as provided by the site"

AshleyT3 commented:
> ... Width of the action dropdowns has shrunk, they are no longer aligned with the other edit boxes in the first variant. ...

That was to propose a different aesthetic since I rarely see word-wrapped static labels (not that there's anything wrong with them). Reverted back to wrapped labels in [88bb6db](https://github.com/pwsafe/pwsafe/pull/1035/commits/88bb6dbb8757f6f12b650596a86e98e6df3a84f2).

> ... "Keyboard Shortcut" has been truncated to "Keyboard" on both variants. ...

That design-time truncation is not part of this PR, exists in 'master', now fixed in [88bb6db](https://github.com/pwsafe/pwsafe/pull/1035/commits/88bb6dbb8757f6f12b650596a86e98e6df3a84f2). No runtime truncation observed in various modes from VGA through to 4K, with various scaling settings. I concur design-time/runtime should match. 

> ..."Double-Click Action:" and "Shift Double-Click Action:" have likewise been truncated in the first variant. ...

Addressed in [88bb6db](https://github.com/pwsafe/pwsafe/pull/1035/commits/88bb6dbb8757f6f12b650596a86e98e6df3a84f2).

A couple of additional items:

* "\n" added to force wrapping at desired word to ensure desired aeshetic. One observed case wrapped at design time (desired), not at runtime (undesired), throwing aesthetics off. 

* A colon appended to "Keyboard shortcut:".

> ... I didn't build and test the tab order, I'd expect the TOTP key field to be tabbed into after keyboard shortcut ...

Yes, key field is after keyboard shortcut. (FYI, you can view the tab order in the dialog editor at design-time by using "CTRL+D." Press ESC when done.)

Side note: Tabbing is currently broken on the "Additional" page given the "Keyboard shortcut" field captures all keystrokes, including tab, disallowing tabbing-through that field. I recommend a tweak (separate backlog issue/PR) to "Keyboard shortcut" for it to ignore tab if it is the first/only key pressed. This would allow tabbing through that field. If other keys are pressed followed by tab, tab would then be captured/used by the "Keyboard shortcut" logic. Let me know if you want a backlog issue opened on this.

> ... Two thoughts on the "Password Safe TOTP Authenticator" text: ...

Changed to "Authenticator Secret".

ronys commented:
Can you resolve the conflicts in src/ui/Windows/res/PasswordSafe3.rc2 and src/ui/Windows/resource3.h? Thanks!

AshleyT3 commented:
> Can you resolve the conflicts in src/ui/Windows/res/PasswordSafe3.rc2 and src/ui/Windows/resource3.h? Thanks!

@ronys done, thanks!

ronys commented on src/ui/Windows/res/PasswordSafe3.rc2:
```diff
@@ -261,7 +261,9 @@ BEGIN
   IDS_CANTSPECIFYMINNUMBER "Sorry, can't specify a minimum number of characters with 'pronounceable' option. Specified values ignored."
-  IDS_OWNSYMBOLSMISSING    "If you specify using your own symbol list, you cannot leave it empty. Unselecting use of symbols."
+  IDS_OWNSYMBOLSMISSING   "If you specify using your own symbol list, you cannot leave it empty. Unselecting use of symbols."
+  IDS_ADDEDITERR_INVALID_TOTP_KEY "Invalid Password Safe Authenticator configuration. %s (%d)"
```
For consistency, should be "Invalid Authenticator Secret Key: %s"
For consistency, I suggest removing the "(%d)", since it's not used elsewhere in PasswordSafe.

ronys commented on src/core/TotpCore.cpp:
```diff
@@ -0,0 +1,101 @@
+
+  std::wstring GetTotpErrorString(TOTP_Result r)
+  {
+    switch (r) {
```
All these strings should be resources.
Also, they should be rephrased to me more actionable by the user. For example, the typical user has no clue as to what 'base32' or 'UTF8' mean. A more useful message would be something like "Key can contain only the following letters: ABCDEFGHIJKLMNOPQRSTUVWXYZ234567 and must be ?? characters long"

ronys commented on src/core/ItemData.cpp:
```diff
@@ -246,13 +274,14 @@ int CItemData::WriteCommon(PWSfile *out) const
   int i;
 
-  const FieldType TextFields[] = {GROUP, TITLE, USER, PASSWORD,
+  const FieldType TextFields[] = {GROUP, TITLE, USER, PASSWORD, TWOFACTORKEY,
```
Would it make more sense to store the key as a binary value, assuming it would be used to calculate the TOTP more often than displaying it?

ronys commented on src/ui/Windows/MainEdit.cpp:
```diff
@@ -2078,16 +2103,99 @@ void DboxMain::UpdateLastClipboardAction(const int iaction)
+  CItemData* pci_credential = &item;
+  if (item.IsDependent()) {
+    pci_credential = GetBaseEntry(&item);
+    ASSERT(pci_credential != NULL);
```
What happens if the user "copies" an entry's auth code, and then deletes the entry?

AshleyT3 commented:
@ronys, regarding...
> * Can you look, reconfirm you want this?
... In updating the "auth code uses password font" change just now, I felt it worthwhile to try making the Display Auth Code (DAC) dialog resizable. If the title bar entry info is truncated, user can resize. If you don't want that, np, just let me know... but seemed worth trying that out.

AshleyT3 commented:
@ronys, For comparison purposes, here's the one with g/t/u in the client area. For clarity, the following is as of older commit [d4d77c3](https://github.com/pwsafe/pwsafe/pull/1035/commits/d4d77c3071f60d4d00c81d1b148e66aa2dd1d5da) so it does not have the password font change... it is only shown here so you can see the client area fields. The latest with caption g/t/u and auth code password font change is shown in prior comment's screenshot.
<img width="800" alt="image" src="https://github.com/pwsafe/pwsafe/assets/32182932/ce66339b-da7c-490b-b095-bf4e4fe51058">

AshleyT3 commented:
@ronys tooltips added to CDisplayAuthDlg. All TOTP commits rebased on top of current master.

ronys commented:
Nice progress! 

A few comments:
1. The Edit control in Auth Code display needs to resize to fit the font:
![image](https://github.com/pwsafe/pwsafe/assets/781072/193f7e62-785e-4338-ba94-cb4b220a5643)
2. I agree that leaving just the group/title/user in chevrons in the title bar is a bit too terse. OTOH, "Password Safe Authenticator" seems wrong to me, as there's no such thing. I Suggest a title bar of "Authentication Code" to be consistent with the terminology, and in the dialog box itself display group/title/user in chevrons.
3. When I copy the auth code by clicking on the pie button in add/edit, it appears that the value in the clipboard isn't updated, as opposed to the behavior when I copy it via the popup menu.

AshleyT3 commented:
@ronys I actually see the DisplayAuthCodeDlg does auto-copy to clipboard after clicking the progress pie (copy) button. 

Can you share your steps, entry info?

Below is a clip of me pasting to Notepad after clicking the Copy button... I see the code change...

https://github.com/pwsafe/pwsafe/assets/32182932/4826934a-8ef8-4d79-b665-46a2be50636f

AshleyT3 commented:
@ronys In case it helps, here are various options for DisplayAuthCodeDlg... based on my above question/concerns, can you choose one of the following or describe a variation?

Option A) The original which uses dialog client area with normal pwsafe (and Windows) field label next to its field control, where title is always visible, user/group only visible if non-empty, the dialog auto-hides/adjusts unused fields of user/group if empty... all the following `A<n>` are part of Option A....


A1: All fields non-empty:

<img width="527" alt="Screenshot 2023-12-08 174526" src="https://github.com/pwsafe/pwsafe/assets/32182932/95ae05b8-0faa-4e4e-8381-fe823e5ec927">


A2: Group is non-empty, username is empty:

<img width="552" alt="Screenshot 2023-12-08 174540" src="https://github.com/pwsafe/pwsafe/assets/32182932/04f41314-d0db-4c0f-aaa0-98973420d03c">


A3: Username is non-empty, group is empty:

<img width="489" alt="Screenshot 2023-12-08 174508" src="https://github.com/pwsafe/pwsafe/assets/32182932/9facd3a5-2209-409c-877f-2917dc5275ec">


A4: Only title is non-empty:

<img width="550" alt="Screenshot 2023-12-08 174552" src="https://github.com/pwsafe/pwsafe/assets/32182932/d4101573-2008-451d-a40b-dbfa9764c7cc">


Option B) Always show group, title, username wrapped in chevrons on separate lines...

![Screenshot 2023-12-08 175153](https://github.com/pwsafe/pwsafe/assets/32182932/746c8d6b-9de4-4c2d-a6ac-041ecdf9ab53)


Option C) Move the caption chevron approach to the client area, so a single line for all 3, always displayed, including empty values...

<img width="524" alt="Screenshot 2023-12-08 180255" src="https://github.com/pwsafe/pwsafe/assets/32182932/e2f31117-a60d-421a-a2f1-ec25bcc47b1f">


Notice above I did not suggest dropping empty values of group/username for chevron approaches. Per my earlier concerns, this is because, without field labels, it's unclear what has been kept/dropped... by itself it's not concretely decipherable to the user lest they seek out the selected entry (or memorize it and just know). I realize it may be obvious in many cases but that's depending on user choice of group names, usernames, etc. It's not a certainty it will be clear. 

So... we can drop empty chevrons but please ACK the nuance pointed out so it's certain what you're specific desired design is here... be great if we can get this next round...

Any thoughts on the above Option A, B, or C, or a variation of your choice?

It's important to clarify up front... an enormous amount of time was spent on the first variation... old school Windows UI out-of-the-box is not as easily sculpted as newer XAML, HTML/CSS and the like... so having a good gist of very specific design upfront can save lots of cycles. I know you want chevron-wrapping approach... I'm going to bat to ensure that's a true final answer, and if so, what is your specific desired aesthetic/UI design?

Much thanks in advance!

AshleyT3 commented:
> Nice progress!
> 
> A few comments:
> 
> 1. The Edit control in Auth Code display needs to resize to fit the font:
>    ![image](https://private-user-images.githubusercontent.com/781072/289026171-193f7e62-785e-4338-ba94-cb4b220a5643.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIwMzE0NDQsIm5iZiI6MTcwMjAzMTE0NCwicGF0aCI6Ii83ODEwNzIvMjg5MDI2MTcxLTE5M2Y3ZTYyLTc4NWUtNDMzOC1iYTk0LWNiNGIyMjBhNTY0My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIwOFQxMDI1NDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iN2E3NGI3YzY3NTlmYjc2NTUxMzk1NDMyZjYyYTUyMzhjZWY1NTNkMDliNWY3Y2ZmMjJmZTViZWVlNWNlNDhhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.QQJ82RBESwY1YokqOWosnRnvb42QXfD2QBq0q8__BPg)

@ronys Great observations, thanks for catching that! The issue was variations given scaling leading to flooring not rounding, fixed w/rounding, will push with next set of changes pending your feedback on other items below.

> 2. I agree that leaving just the group/title/user in chevrons in the title bar is a bit too terse. OTOH, "Password Safe Authenticator" seems wrong to me, as there's no such thing. I Suggest a title bar of "Authentication Code" to be consistent with the terminology, and in the dialog box itself display group/title/user in chevrons.

ACK re changing DisplayAuthCodeDlg title bar caption to "Authentication Code". 

Regarding chevrons... I'd like to ensure that's really what you want, and whether you want that or alternatives I suggest, I'd like to get from you the gist of your specific design direction (as there are a few directions with both approaches). See discussion/choices further below.

> 3. When I copy the auth code by clicking on the pie button in add/edit, it appears that the value in the clipboard isn't updated, as opposed to the behavior when I copy it via the popup menu.

I do not see that behavior. You really do not see anything on the clipboard change? When I click that button, the clipboard is continuously updated.

Are you sure you are not copying something else to the clipboard, or something like that? (I assume not but just in case, asking.)

Can you share specifics such as entry information of entry you are copying from... is it an alias or shortcut? How you are verifying? 

See the following video clip for my experience. You can see me pasting to Notepad over and over where the pasted code changes can be seen... 

https://github.com/pwsafe/pwsafe/assets/32182932/254952e1-5f8d-48bc-a9a9-cf1dabb6c991

---

Group/Title/Username UI design discussion: Following are various options for DisplayAuthCodeDlg... based on my above question/concerns, can you choose one of the following or describe a preferred variation not shown?

Option A) The original which uses dialog client area with normal pwsafe (and Windows) field label next to its field control, where title is always visible, user/group only visible if non-empty, the dialog auto-hides/adjusts unused fields of user/group if empty... all the following `A<n>` are part of Option A....


Sub-option A1: All fields non-empty:

<img width="527" alt="Screenshot 2023-12-08 174526" src="https://github.com/pwsafe/pwsafe/assets/32182932/95ae05b8-0faa-4e4e-8381-fe823e5ec927">


Sub-option A2: Group is non-empty, username is empty:

<img width="552" alt="Screenshot 2023-12-08 174540" src="https://github.com/pwsafe/pwsafe/assets/32182932/04f41314-d0db-4c0f-aaa0-98973420d03c">


Sub-option A3: Username is non-empty, group is empty:

<img width="489" alt="Screenshot 2023-12-08 174508" src="https://github.com/pwsafe/pwsafe/assets/32182932/9facd3a5-2209-409c-877f-2917dc5275ec">


Sub-option A4: Only title is non-empty:

<img width="550" alt="Screenshot 2023-12-08 174552" src="https://github.com/pwsafe/pwsafe/assets/32182932/d4101573-2008-451d-a40b-dbfa9764c7cc">


Option B) Always show group, title, username wrapped in chevrons on separate lines...

![Screenshot 2023-12-08 175153](https://github.com/pwsafe/pwsafe/assets/32182932/746c8d6b-9de4-4c2d-a6ac-041ecdf9ab53)


Option C) Move the caption chevron approach to the client area, so a single line for all 3, always displayed, including empty values...

<img width="524" alt="Screenshot 2023-12-08 180255" src="https://github.com/pwsafe/pwsafe/assets/32182932/e2f31117-a60d-421a-a2f1-ec25bcc47b1f">


Notice above I did not suggest dropping empty values of group/username for chevron approaches. This is because, without field labels, it's unclear what has been kept/dropped... by itself it's not concretely decipherable to the user lest they seek out the selected entry (or memorize it and just know). I realize it may be obvious in many cases but that's depending on user choice of group names, usernames, etc. Clarity is not a certainty.

So... we can drop empty chevrons but please ACK the nuance pointed out so it's certain what your specific desired design is... it would be great if we can get this this round...

Any thoughts on the above Option A, B, or C, or a variation of your choice?

You know... but old school Windows UI out-of-the-box is not as easily sculpted as newer XAML, HTML/CSS and the like... so having a good gist of your specific design upfront can save lots of cycles.  The first round with dynamic field adjustments took a bit of time... no biggie... but for the next round, given each avenue itself has a number of variations, it would be great to get your detailed thoughts on this beyond the notion of using chevrons in the client area, if that makes sense. 

Hope this one summary helps... much thanks in advance!

ronys commented:
Re the clipboard not refreshing: This happens only when I click on the pie button *in the add/edit dialog box*.

ronys commented:
Re the UI options: "C" works best for me, as it's the most consistent with the group/title/user display already used. Also, please change the title bar for "Password Safe Authenticator" to "Authenication Code".

AshleyT3 commented:
@ronys PR updated.

AshleyT3 commented:
@ronys For PR update, all checks passed except Ubuntu-latest for apparent infra issues, can you look when you have a sec? I will hold off repushing to trigger... not required for you to review.

All passing exception Ubuntu-latest due to apparent infra issues:
<img width="652" alt="image" src="https://github.com/pwsafe/pwsafe/assets/32182932/782e8baa-a993-4cb0-be90-bae8c8663ce3">

Ubuntu-latest error observed:

```
...
Err:9 mirror+file:/etc/apt/apt-mirrors.txt jammy-updates/main amd64 libcurl4-openssl-dev amd64 7.81.0-1ubuntu1.14
  404  Not Found [IP: 52.252.163.49 80]
...
E: Failed to fetch mirror+file:/etc/apt/apt-mirrors.txt/pool/main/c/curl/libcurl4-openssl-dev_7.81.0-1ubuntu1.14_amd64.deb  404  Not Found [IP: 52.252.163.49 80]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
Fetched 19.1 MB in 2s (92[98](https://github.com/pwsafe/pwsafe/actions/runs/7152900254/job/19478736485?pr=1035#step:3:99) kB/s)
Error: Process completed with exit code 100.
```

Screenshot from log:

<img width="837" alt="image" src="https://github.com/pwsafe/pwsafe/assets/32182932/871d158d-c0e1-47cd-8532-cb6a6711a9ef">

AshleyT3 commented:
@ronys for when you return, same pwsafe build check `CMake Ubuntu-latest...` is failing due to an inaccessible apt mirror at 52.252.75.106 in efforts to get `libcurl4-openssl-dev amd64 7.81.0-1ubuntu1.14` ... apparently affecting all currently open PRs.

No changes to this PR beyond a fresh rebased on top of master... it has your requested changes for review.

ronys commented:
Hi,

The Ubuntu build failure is annoying, but seems to be a GitHub issue, and not something I can fix locally.

As to the TOTP feature, I think it's almost ready for merge. The only thing I see missing is the Two Factor Secret field in the list of fields in the "Advanced" export options.

AshleyT3 commented:
> ...
> As to the TOTP feature, I think it's almost ready for merge. The only thing I see missing is the Two Factor Secret field in the list of fields in the "Advanced" export options.

Thanks @ronys, two factor key added to import/export/sync/compare dialogs in commit [da3d7cb](https://github.com/pwsafe/pwsafe/pull/1035/commits/da3d7cb9f4e4a0f93fb8284d30265d80dccb6ee6).

ronys commented:
The TOTP field exclusion for text export works fine. For XML, though, it's detected but not enforced :

     Export of data was restricted to certain fields by the user.
     The user specified that the following fields be excluded from the export:
     Two Factor Key, TOTP Config, TOTP Start Time, TOTP Time Step, TOTP Length, Created Time

`<!-- 
     Export of data was restricted to certain fields by the user.
     The user specified that the following fields be excluded from the export:
     Two Factor Key, TOTP Config, TOTP Start Time, TOTP Time Step, TOTP Length, Created Time
 --> 

	<entry id="1">
		<group><![CDATA[totp]]></group>
		<title><![CDATA[t1]]></title>
		<password><![CDATA[&vJv=YfjZZ:H]]></password>
		<twofactorkey><![CDATA[aabbccddeeffeedd]]></twofactorkey>
		<uuid><![CDATA[338b5c1a76f6483180f235bea8d60915]]></uuid>
		<rmtimex>2023-11-28T19:43:53</rmtimex>
	</entry>`

ronys commented:
Also, it would be useful, I think, to have the Authentication Secret  field in the Display Filter rules, I' d add it with the same criteria as "Group", that is:
- is present
- is not present
- equals
- does not equal
- begins with
- does not begin with
- ends with
- does not end with
- contains
- does not contain
- contains any of
- does not contain any of
- contains all of
- does not contain all of

ronys commented:
For consistency, the field should appear as "Authentication Secret" in user visible fields (Export and display filters), as this is how it's named in the add/edit dialog box.
"Two Factor Key" could be confused with Yubikey, which is documented as "Two factor authentication" for protecting the PasswordSafe database.

AshleyT3 commented:
> The TOTP field exclusion for text export works fine. For XML, though, it's detected but not enforced :
> 
> ```
>  Export of data was restricted to certain fields by the user.
>  The user specified that the following fields be excluded from the export:
>  Two Factor Key, TOTP Config, TOTP Start Time, TOTP Time Step, TOTP Length, Created Time
> ```
> 
> `
> 
> ```
> <entry id="1">
> 	<group><![CDATA[totp]]></group>
> 	<title><![CDATA[t1]]></title>
> 	<password><![CDATA[&vJv=YfjZZ:H]]></password>
> 	<twofactorkey><![CDATA[aabbccddeeffeedd]]></twofactorkey>
> 	<uuid><![CDATA[338b5c1a76f6483180f235bea8d60915]]></uuid>
> 	<rmtimex>2023-11-28T19:43:53</rmtimex>
> </entry>`
> ```



> Also, it would be useful, I think, to have the Authentication Secret field in the Display Filter rules, I' d add it with the same criteria as "Group", that is:
> 
> * is present
> * is not present
> * equals
> * does not equal
> * begins with
> * does not begin with
> * ends with
> * does not end with
> * contains
> * does not contain
> * contains any of
> * does not contain any of
> * contains all of
> * does not contain all of

> For consistency, the field should appear as "Authentication Secret" in user visible fields (Export and display filters), as this is how it's named in the add/edit dialog box. "Two Factor Key" could be confused with Yubikey, which is documented as "Two factor authentication" for protecting the PasswordSafe database.

@ronys Thanks... your above feedback items are addressed in commits [4b9037f](https://github.com/pwsafe/pwsafe/pull/1035/commits/4b9037ff3a2401e40bf0207e90a621e79702d8cd), [2058ff7](https://github.com/pwsafe/pwsafe/pull/1035/commits/2058ff70b96a8c0b636af80ec90ba19bf24d5f1e), and [d738cd6](https://github.com/pwsafe/pwsafe/pull/1035/commits/d738cd61d258e67ee12beac8078a950fe12e42bc).

ronys reviewed (approved):
Looks good, including comparing generated TOTPs against external references, both 64 and 32 bit builds.
Time to merge into master, continue documentation & i18n from there.

Huge thanks!

AshleyT3 commented:
> Looks good, including comparing generated TOTPs against external references, both 64 and 32 bit builds. Time to merge into master, continue documentation & i18n from there.
> 
> Huge thanks!
...
> @[ronys](https://github.com/ronys) ronys merged commit [2c765d1](https://github.com/pwsafe/pwsafe/commit/2c765d1d3be14fabd92577db30591f0012a23ad4) into pwsafe:master [3 hours ago](https://github.com/pwsafe/pwsafe/pull/1035#event-11269155588)
5 checks passed

@ronys Wow... that is great news... thank you for the checks, the merge... and all collaboration leading to it! On a number of levels, super awesome.
